shared_model/interfaces/common_objects/impl/amount.cpp

```diff
@@ -0,0 +1,71 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "interfaces/common_objects/amount.hpp"
+
+#include <regex>
+
+#include <boost/multiprecision/cpp_dec_float.hpp>
+#include "utils/string_builder.hpp"
+
+namespace shared_model {
+  namespace interface {
+    Amount::Amount(const std::string &amount) : Amount(std::string(amount)) {}
+    Amount::Amount(std::string &&amount)
+        : amount_(std::move(amount)),
+          multiprecision_repr_([this] {
+            static const std::regex r("([0-9]+)(\\.([0-9]+))?");
+            // 123.456 will have the following groups:
+            //   [0] -> 123.456, [1] -> 123
+            //   [2] -> .456,    [3] -> 456
+            std::smatch match;
+            if (std::regex_match(this->amount_, match, r)
+                && match.size() == 4) {
+              this->precision_ = match[3].length();
+              auto str = match[0].str();
+              decltype(str.size()) pos = match[1].length();
```

Comment:
auto?

Reply:
for some reason it deduce signed type which leads to warning :(

Reply:
size_t maybe?

Reply:
yeah, it can be